label_properties now generates a source location if there is none

kroening opened the pull request:
A `source_locationt` with `id()` `nil` signals that the source location is not
given; an empty string signals that the source location is given.  Goto
instructions without source location are perfectly valid, but we need a
source location to carry the property id.  This fixes `label_properties` to
create a source location with the right function name if none is given.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented on src/goto-programs/set_properties.cpp:
```diff
@@ -92,7 +101,7 @@ void label_properties(
 void label_properties(goto_programt &goto_program)
 {
   std::map<irep_idt, std::size_t> property_counters;
-  label_properties(goto_program, property_counters);
+  label_properties({}, goto_program, property_counters);
 }
```
Can we just remove this function?

kroening replied:
Done

tautschnig replied:
Hmm, that wasn't actually unused, was it? It seems JBMC now fails to build.

kroening replied:
Now added a variant for JBMC.

tautschnig reviewed (approved):
In a way I'm glad this is now failing one test, for that demonstrates that the change makes a difference :-) (Still needs to be fixed, though...)

kroening committed:
label_properties now generates a source location if there is none

A source_locationt with id() nil signals that the source location is not
given; an empty string signals that the source location is given.  Goto
instructions without source location are perfectly valid, but we need a
source location to carry the property id.  This fixes label_properties to
create a source location with the right function name if none is given.

kroening commented:
> In a way I'm glad this is now failing one test, for that demonstrates that the change makes a difference :-) (Still needs to be fixed, though...)

I've got no idea why the contract checking fails here -- who's the right person to ask?

tautschnig committed:
Code contracts: add source location to function pointer contracts

The expression constructed in the parser never got a source location
assigned, yet it was this sub-expression that was actually retained by
type checking. Consequently, properties were printed without a source
location.

tautschnig commented:
> > In a way I'm glad this is now failing one test, for that demonstrates that the change makes a difference :-) (Still needs to be fixed, though...)
> 
> I've got no idea why the contract checking fails here -- who's the right person to ask?

I have pushed a fix, this was a bug introduced in 63cd722f.

kroening commented:
> I have pushed a fix, this was a bug introduced in [63cd722](https://github.com/diffblue/cbmc/commit/63cd722f266dc63ec0e952bdc111e7ddf6ab014f).

Many thanks!
